aptos-move/framework/aptos-framework/sources/object_code_deployment.move

```diff
@@ -148,4 +160,63 @@ module aptos_framework::object_code_deployment {
 
         event::emit(Freeze { object_address: object::object_address(&code_object), });
     }
+
+    /// Registers a capability proof for the `code_object` to allow generating the signer for the `code_object` later via
+    /// `object_code_deployment::generate_signer`.
+    ///
+    /// This can only be called by the owner of the `code_object` or the package itself.
+    public entry fun register_signer_capability_proof<ProofType>(
+        owner_or_package: &signer
+    ) acquires CodeSignerCapability, ManagingRefs {
+        code::check_code_publishing_permission(owner_or_package);
+
+        let proof_type = type_info::type_of<ProofType>();
+        let code_object_address = proof_type.account_address();
+        // Disallow registering a capability proof for an object that is not a code object.
+        assert_is_code_object(code_object_address);
+
+        let caller_addr = signer::address_of(owner_or_package);
+        let is_code_object_owner =
+            object::is_owner(object::address_to_object<PackageRegistry>(code_object_address), caller_addr);
+        let is_package_itself = caller_addr == code_object_address;
+        assert!(
+            is_code_object_owner || is_package_itself,
+            error::permission_denied(ENOT_CODE_OBJECT_OWNER),
+        );
+
+        if (!exists<CodeSignerCapability>(code_object_address)) {
+            let code_object_signer = &object::generate_signer_for_extending(&ManagingRefs[code_object_address].extend_ref);
+            move_to(code_object_signer, CodeSignerCapability { capability_proof: proof_type });
+        } else {
+            CodeSignerCapability[code_object_address].capability_proof = proof_type;
+        };
+    }
+
+    /// Generates a signer for the `code_object` if the caller has registered a capability proof for it.
+    public fun generate_signer<ProofType>(
+        _proof: &ProofType
+    ): signer acquires CodeSignerCapability, ManagingRefs {
+        let proof_type = type_info::type_of<ProofType>();
+        let code_object_address = proof_type.account_address();
+        // This is redundant with the check in `register_signer_capability_proof`, but we want to cautious here and also
+        // fail early if the `code_object` is not a code object.
+        assert_is_code_object(code_object_address);
+
+        assert!(exists<CodeSignerCapability>(code_object_address), error::not_found(ENO_SIGNER_CAPABILITY_CONFIGURED));
+        let proof_required = CodeSignerCapability[code_object_address].capability_proof;
+        assert!(proof_type == proof_required, error::permission_denied(ENO_CODE_PERMISSION));
+
+        object::generate_signer_for_extending(&ManagingRefs[code_object_address].extend_ref)
```

Comment:
Interesting, this is so you can have different proofs later?

Reply:
Yes the code object owner can change it later